hcl/lex/lex.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lex
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/mineiros-io/terramate/errors"
+)
+
+// TokenIdent creates a new identifier token
+func TokenIdent(name string) *hclwrite.Token {
+	return &hclwrite.Token{
+		Type:  hclsyntax.TokenIdent,
+		Bytes: []byte(name),
+	}
+}
+
+// TokenOBrace creates a new { token.
+func TokenOBrace() *hclwrite.Token {
+	return &hclwrite.Token{
+		Type:  hclsyntax.TokenOBrace,
+		Bytes: []byte{byte(hclsyntax.TokenOBrace)},
+	}
+}
+
+// TokenEqual creates a new = token.
+func TokenEqual() *hclwrite.Token {
+	return &hclwrite.Token{
+		Type:  hclsyntax.TokenEqual,
+		Bytes: []byte{byte(hclsyntax.TokenEqual)},
+	}
+}
+
+// TokenEquals compare if two tokens have the same type and bytes.
+func TokenEquals(token1, token2 *hclwrite.Token) bool {
+	return token1.Type == token2.Type &&
+		string(token1.Bytes) == string(token2.Bytes)
+}
+
+// FindTokenSequence finds the first match of the given token sequence
+// on the given tokens, returning the position of the last token of the
+// matched sequence, so the next token (if any) would be what is after the
+// given sequence. The contents of the tokens are also matched, not only their type.
+func FindTokenSequence(tokens hclwrite.Tokens, seq ...*hclwrite.Token) (int, bool) {
```

Comment:
For me the design of the algorithm here and the API are quite confusing...
Why we don't treat like a FindSubst() kind of function? that returns the position of first element that matched the whole sequence? Then the caller just knows it needs to skip the size of seq. Example:

```
    begin, ok := FindTokenSequence(tokens, stack, obrace)
    if !ok { }
    stackTokens := tokens[begin+2:] // or tokens[begin+len(seq)]
```
The above API is more useful for other parser tasks.

Regarding the algorithm, it looks more complex than needed, no? Or maybe I'm missing something... The `matchHead`, `pendingMatches` and comparing `TokenEquals()` in both loops seem unneeded for a substr-like function.

What about something like that?

```
func FindTokenSequence(tokens hclwrite.Tokens, seq ...*hclwrite.Token) (int, bool) {
	if len(seq) == 0 {
		return 0, false
	}
outer:
	for i := 0; i < len(tokens); i++ {
		found := false
		for j := 0; j < len(seq); j++ {
			if !TokenEquals(tokens[i+j], seq[j]) {
				continue outer
			}
			found = true
		}
		if found {
			return i + len(seq) - 1, true
		}
	}
	return 0, false
}
```
even though I'd rather return `i` (the position of the first matched token) instead of last... The algorithm above can be way simpler with the other API (I guess).

but no strong feeling, just found it a little odd/different than what I was expecting.
Feel free to ignore, or maybe improve later.

Reply:
Yeah I had a feeling that it was way too complicated at the time but I was trying to figure out a bunch of different stuff and was not able to come up with the simpler solution and just got happy as soon as it worked :laughing: , will adjust :+1:

Reply:
Told you multiple times I don't have much experience with parsers :laughing:

Reply:
Will just adjust the code a little, I think it can be even simpler and the current code panics when acessing tokens[i+j] (it can be out of bounds on some cases)